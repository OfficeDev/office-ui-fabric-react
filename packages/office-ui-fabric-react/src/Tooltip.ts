export * from './packages/react-surfaces/Tooltip';
