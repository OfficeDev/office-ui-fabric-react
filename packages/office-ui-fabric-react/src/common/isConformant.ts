// eslint-disable-next-line import/no-extraneous-dependencies -- this file is for testing
import { isConformant as isConformantBase, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant(testInfo: IsConformantOptions) {
  const defaultOptions = {
    disabledTests: ['has-docblock'],
  };

  isConformantBase(defaultOptions, testInfo);
}
