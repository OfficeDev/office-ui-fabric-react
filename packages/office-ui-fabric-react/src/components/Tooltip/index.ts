export * from './Tooltip';
export * from './Tooltip.types';
export * from './TooltipHost';
export * from './TooltipHost.types';
export * from './TooltipHostDeferred';
export * from '../../common/DirectionalHint';
