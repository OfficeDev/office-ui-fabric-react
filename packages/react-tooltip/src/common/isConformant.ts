import { isConformant as baseIsConformant, IsConformantOptions } from '@fluentui/react-conformance';

export function isConformant<TProps>(testInfo: Omit<IsConformantOptions<TProps>, 'componentPath'>) {
  const defaultOptions = {
    asPropHandlesRef: true,
    componentPath: module!.parent!.filename.replace('.test', ''),
  };

  baseIsConformant<TProps>(defaultOptions, testInfo);
}
