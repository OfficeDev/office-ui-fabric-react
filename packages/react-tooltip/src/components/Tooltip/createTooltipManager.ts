import { TooltipManager, TriggerTooltipArgs } from '@fluentui/react-shared-contexts';

/**
 * Create an instance of TooltipManager
 */
export const createTooltipManager: () => TooltipManager = () => {
  let timeoutId: ReturnType<typeof setTimeout> | undefined = undefined;

  const setTooltipTimeout = (fn: () => void, delay: number) => {
    clearTooltipTimeout();
    timeoutId = setTimeout(fn, delay);
  };

  const clearTooltipTimeout = () => {
    if (timeoutId !== undefined) {
      clearTimeout(timeoutId);
      timeoutId = undefined;
    }
  };

  // The trigger element that is either hovered or focused
  let activeTrigger: HTMLElement | undefined = undefined;

  // The args for the tooltip that is currently visible
  let visible: TriggerTooltipArgs | undefined = undefined;

  const notifyEnterTrigger = (args: TriggerTooltipArgs) => {
    activeTrigger = args.trigger;

    if (visible?.trigger === args.trigger) {
      // The tooltip is already visible; cancel the hide timer if it was running
      clearTooltipTimeout();
    } else {
      // Show the new tooltip immediately if there is already one visible
      const showDelay = visible ? 0 : args.showDelay;

      setTooltipTimeout(() => {
        // Double check that the trigger is still being hovered
        if (activeTrigger === args.trigger) {
          hideAll();
          args.setVisible(true);
          visible = args;
        }
      }, showDelay);
    }
  };

  const notifyLeaveTrigger = (trigger: HTMLElement) => {
    // Ignore if this is coming from an element that's not the trigger
    if (activeTrigger !== trigger) {
      return;
    }

    activeTrigger = undefined;
    if (!visible) {
      // The tooltip is already hidden; cancel the show timer if it was running
      clearTooltipTimeout();
    } else {
      setTooltipTimeout(() => {
        // Double check that the trigger or tooltip didn't get hovered in the meantime
        if (!activeTrigger || activeTrigger !== visible?.trigger) {
          hideAll();
        }
      }, visible.hideDelay);
    }
  };

  const notifyEnterTooltip = () => {
    // Treat hovering over the tooltip the same as hovering over the tooltip's trigger element
    if (visible) {
      activeTrigger = visible.trigger;
    }
  };

  const notifyLeaveTooltip = () => {
    if (visible) {
      notifyLeaveTrigger(visible.trigger);
    }
  };

  const hideAll = () => {
    clearTooltipTimeout();
    visible?.setVisible(false);
    visible = undefined;
  };

  return {
    notifyEnterTrigger,
    notifyLeaveTrigger,
    notifyEnterTooltip,
    notifyLeaveTooltip,
    hideAll,
  };
};
