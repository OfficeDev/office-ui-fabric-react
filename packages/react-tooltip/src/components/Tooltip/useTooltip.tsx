import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useId, useIsSSR, useMergedRefs } from '@fluentui/react-utilities';
import { usePopper } from '@fluentui/react-positioning';
import { useTheme } from '@fluentui/react-theme-provider';
import { TooltipProps, TooltipShorthandProps, TooltipState, TooltipTriggerProps } from './Tooltip.types';
import { arrowHeight, tooltipBorderRadius } from './useTooltipStyles';
import { useTooltipContext } from '../TooltipProvider';
import { createTooltipManager } from '../../TooltipManager';

/**
 * Names of the shorthand properties in TooltipProps
 * {@docCategory Tooltip}
 */
export const tooltipShorthandProps: TooltipShorthandProps[] = ['content'];

const mergeProps = makeMergeProps<TooltipState>({ deepMerge: tooltipShorthandProps });

const mergeCallbacks = <Event,>(
  callback1: ((ev: Event) => void) | undefined,
  callback2: ((ev: Event) => void) | undefined,
) => {
  if (!callback1 || !callback2) {
    return callback1 || callback2;
  }

  return (ev: Event) => {
    callback1(ev);
    callback2(ev);
  };
};

/**
 * Create the state required to render Tooltip.
 *
 * The returned state can be modified with hooks such as useTooltipStyles,
 * before being passed to renderTooltip.
 *
 * @param props - props from this instance of Tooltip
 * @param ref - reference to root HTMLElement of Tooltip
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory Tooltip}
 */
export const useTooltip = (props: TooltipProps, defaultProps?: TooltipProps): TooltipState => {
  const [visible, setVisible] = React.useState(false);

  const state = mergeProps(
    {
      children: props.children,
      content: {
        as: 'div',
        role: 'tooltip',
        id: useId('tooltip-'),
      },
      position: 'above',
      align: 'center',
      offset: 4,
      showDelay: 250,
      hideDelay: 250,
      visible,
      isContentRendered: !useIsSSR(),
    },
    defaultProps && resolveShorthandProps(defaultProps, tooltipShorthandProps),
    resolveShorthandProps(props, tooltipShorthandProps),
  );

  const theme = useTheme();

  const popper = usePopper({
    enabled: state.visible,
    position: state.position,
    align: state.align,
    offset: [0, state.offset + (state.noArrow ? 0 : arrowHeight)],
    arrowPadding: theme?.global ? 2 * parseInt(tooltipBorderRadius(theme), 10) : 0,
  });

  state.content.ref = useMergedRefs(state.content.ref, popper.containerRef);
  state.arrowRef = popper.arrowRef;

  const { tooltipManagerRef } = useTooltipContext();

  if (!tooltipManagerRef.current) {
    tooltipManagerRef.current = createTooltipManager();
  }

  state.content.onPointerEnter = mergeCallbacks(
    () => tooltipManagerRef.current?.notifyEnterTooltip(),
    state.content.onPointerEnter,
  );
  state.content.onPointerLeave = mergeCallbacks(
    () => tooltipManagerRef.current?.notifyLeaveTooltip(),
    state.content.onPointerLeave,
  );

  const onEnterTrigger = (ev: React.SyntheticEvent<HTMLElement>) => {
    const tgt = state.targetRef?.current ?? ev.currentTarget;
    popper.targetRef.current = tgt;

    // For tooltips that only show when truncated, don't show if the target's client size <= scroll size
    if (state.onlyIfTruncated && tgt.clientWidth >= tgt.scrollWidth && tgt.clientHeight >= tgt.scrollHeight) {
      return;
    }

    tooltipManagerRef.current?.notifyEnterTrigger({
      setVisible,
      trigger: ev.currentTarget,
      showDelay: state.showDelay,
      hideDelay: state.hideDelay,
    });
  };

  const onLeaveTrigger = (ev: React.SyntheticEvent<HTMLElement>) => {
    tooltipManagerRef.current?.notifyLeaveTrigger(ev.currentTarget);
  };

  const triggerAriaProps: TooltipTriggerProps = {};

  if (state.type === 'description') {
    if (state.isContentRendered) {
      triggerAriaProps['aria-describedby'] = state.content.id;
    }
  } else if (typeof state.content.children !== 'string') {
    if (state.isContentRendered) {
      triggerAriaProps['aria-labelledby'] = state.content.id;
    }
  } else {
    // If the content is the trigger's label, and it is a simple string, then we can use the aria-label prop, and
    // we don't need to render the tooltip content if it isn't visible
    triggerAriaProps['aria-label'] = state.content.children as string;
    if (!state.visible) {
      state.isContentRendered = false;
    }
  }

  if (typeof state.children === 'function') {
    // If a render function was passed in as the child, pass the props to the function
    state.children = state.children({
      onPointerEnter: onEnterTrigger,
      onPointerLeave: onLeaveTrigger,
      onFocus: onEnterTrigger,
      onBlur: onLeaveTrigger,
      ...triggerAriaProps,
    }) as TooltipState['children'];
  } else {
    const child = React.Children.only(state.children);
    if (child.type !== React.Fragment) {
      // Attach the extra props by cloning the child
      state.children = React.cloneElement(child, {
        onPointerEnter: mergeCallbacks(child.props.onPointerEnter, onEnterTrigger),
        onPointerLeave: mergeCallbacks(child.props.onPointerLeave, onLeaveTrigger),
        onFocus: mergeCallbacks(child.props.onFocus, onEnterTrigger),
        onBlur: mergeCallbacks(child.props.onBlur, onLeaveTrigger),
        ...triggerAriaProps,
      });
    } else if (process.env.NODE_ENV !== 'production') {
      throw new Error(
        'TooltipTrigger has unsupported children. It can only contain a single' +
          ' React element (not a fragment), or a render function.',
      );
    }
  }

  return state;
};
