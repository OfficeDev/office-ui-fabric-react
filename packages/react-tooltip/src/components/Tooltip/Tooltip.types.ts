import * as React from 'react';
import { ComponentState } from '@fluentui/react-utilities';
import { TooltipProps } from '../../common/types';

export { TooltipProps };

/**
 * Names of the shorthand properties in TooltipProps
 * {@docCategory Tooltip}
 */
export type TooltipShorthandProps = 'arrow';

/**
 * Names of TooltipProps that have a default value in useTooltip
 * {@docCategory Tooltip}
 */
export type TooltipDefaultedProps = 'position' | 'align' | 'offset';

/**
 * {@docCategory Tooltip}
 */
export type TooltipState = ComponentState<
  React.Ref<HTMLElement>,
  TooltipProps & {
    visible: boolean;
  },
  TooltipShorthandProps,
  TooltipDefaultedProps
>;
