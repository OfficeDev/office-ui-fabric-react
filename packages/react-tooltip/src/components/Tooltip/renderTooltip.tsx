import * as React from 'react';
import { Portal } from '@fluentui/react-portal';
import { getSlots } from '@fluentui/react-utilities';
import { TooltipState } from './Tooltip.types';
import { tooltipShorthandProps } from './useTooltip';

/**
 * Render the final JSX of Tooltip
 * {@docCategory Tooltip}
 */
export const renderTooltip = (state: TooltipState) => {
  const { slots, slotProps } = getSlots(state, tooltipShorthandProps);

  const { children, ...restOfContentProps } = slotProps.content;

  return (
    <>
      {state.children}
      {state.isContentRendered && (
        <Portal>
          <slots.content {...restOfContentProps}>
            {!state.noArrow && <div ref={state.arrowRef} className={state.arrowClassName} />}
            {children}
          </slots.content>
        </Portal>
      )}
    </>
  );
};
