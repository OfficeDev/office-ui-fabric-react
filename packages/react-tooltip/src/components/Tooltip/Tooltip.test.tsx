import * as React from 'react';
import { Tooltip } from './Tooltip';
import { isConformant } from '../../common/isConformant';
import { render, RenderResult } from '@testing-library/react';

describe('Tooltip', () => {
  isConformant({
    Component: Tooltip,
    displayName: 'Tooltip',
    requiredProps: { content: 'Example', children: <button /> },
    disabledTests: [
      // Tooltip renders into a Portal, which confuses these tests
      'component-handles-ref',
      'component-has-root-ref',
      'component-handles-classname',
      'as-renders-fc',
      'as-passes-as-value',
      'as-renders-html',
    ],
  });

  let wrapper: RenderResult | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  it('renders only aria-label for a simple label tooltip', () => {
    const tooltipText = 'The tooltip text';
    wrapper = render(
      <Tooltip content={tooltipText}>
        <button data-testid="the-target" />
      </Tooltip>,
    );

    const target = wrapper.getByTestId('the-target');
    expect(target.getAttribute('aria-label')).toBe(tooltipText);

    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('renders the content of a nontrivial tooltip', () => {
    const tooltipId = 'the-tooltip-id';
    wrapper = render(
      <Tooltip
        id={tooltipId}
        data-testid="the-tooltip"
        content={
          <span>
            This is a <strong>formatted</strong> tooltip
          </span>
        }
      >
        <button data-testid="the-target" />
      </Tooltip>,
    );

    const tooltip = wrapper.getByTestId('the-tooltip');
    const target = wrapper.getByTestId('the-target');
    expect(tooltip.id).toBe(tooltipId);
    expect(target.getAttribute('aria-labelledby')).toBe(tooltipId);

    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('renders a description tooltip content always', () => {
    wrapper = render(
      <Tooltip content="Description tooltip" type="description" data-testid="the-tooltip">
        <button data-testid="the-target" />
      </Tooltip>,
    );

    const tooltip = wrapper.getByTestId('the-tooltip');
    const target = wrapper.getByTestId('the-target');
    expect(target.getAttribute('aria-describedby')).toBe(tooltip.id);

    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
