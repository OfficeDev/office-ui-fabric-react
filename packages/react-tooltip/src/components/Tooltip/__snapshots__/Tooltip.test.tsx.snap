// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders a default state 1`] = `
<div
  className=""
  role="tooltip"
>
  <div
    className="tooltip-arrow"
  >
    
  </div>
  Default Tooltip
</div>
`;
